numpy/lib/npyio.py

```diff
@@ -181,6 +181,9 @@ class NpzFile(Mapping):
 
     def __init__(self, fid, own_fid=False, allow_pickle=False,
                  pickle_kwargs=None):
+        # Make __exit__ safe if zipfile_factory raises an exception
+        self.zip = None
+        self.fid = None
```

Comment:
It's tempting to set these outside `__init__`, so that even a `Ctrl+C` before these two lines will not cause an exception in `__del__`

Reply:
@eric-wieser correct me if I am wrong, but we cannot move this outside `__init__` right. That would make these as static variables, which we probably don't want.

I was thinking a a fix could be to use `hasattr` in `def close` before accessing `self.zip` and `self.fid`. Similar to what is being done is `_datasource.py` for class `DataSource`

Reply:
@eric-wieser I'm not sure what you're asking for.  Python doesn't allow this for example:
```python
>>> class Foo:
...     self.x = 444
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in Foo
NameError: name 'self' is not defined
```
So what do you mean by "set these outside init"?  Set them where?

Reply:
They could be defined as class variables rather than in the constructor, to take your example:

```python
class Foo:
    x = 444
```

Reply:
@rossbar Thanks, amended and force-pushed.